src/main/java/hexgui/sgf/SgfReader.java

```diff
@@ -0,0 +1,411 @@
+        node.addSetup(HexColor.EMPTY, parsePoint(val));
+      } else if (name.equals("LB")) {
+        node.addLabel(val);
+      } else if (name.equals("FF")) {
+        node.setSgfProperty(name, val);
+        x = parseInt(val);
+        if (x < 1 || x > 4) throw sgfError("Invalid SGF Version! (" + x + ")");
+      } else if (name.equals("GM")) {
+        node.setSgfProperty(name, val);
+        if (!isroot) sgfWarning("GM property in non-root node!");
+        if (parseInt(val) != GM_HEXGAME) throw sgfError("Not a Hex game!");
+      } else if (name.equals("SZ")) {
+        node.setSgfProperty(name, val);
+        if (!isroot) sgfWarning("GM property in non-root node!");
+        Dimension dim = new Dimension();
+        String sp[] = val.split(":");
+        if (sp.length == 1) {
+          x = parseInt(sp[0]);
+          dim.setSize(x, x);
+        } else if (sp.length == 2) {
+          x = parseInt(sp[0]);
+          y = parseInt(sp[1]);
+          dim.setSize(x, y);
+        } else {
+          throw sgfError("Malformed boardsize!");
+        }
+        m_gameinfo.setBoardSize(dim);
+      } else if (name.equals("AP")) {
+        node.setSgfProperty(name, val);
+        String regex = "HexGui:0\\.[0-9](\\z|[^0-9].*)";
+        if (val.matches(regex)) {
+          // version HexGui:0.9 or earlier
+          m_swap_bug = true;
+        }
+      } else {
+        node.setSgfProperty(name, val);
+      }
+    }
+  }
+
+  // Parse an SGF "SimpleText" property value.
+  private String parseValue() throws SgfError, IOException {
+    int ttype = m_tokenizer.nextToken();
+    if (ttype != '[') {
+      throw sgfError("Property missing opening '['.");
+    }
+
+    StringBuilder sb = new StringBuilder(256);
+    boolean quoted = false;
+    while (true) {
+      int ch = m_reader.read();
+      if (ch < 0) {
+        throw sgfError("Property runs to EOF.");
+      }
+      // Don't rely on the default conversion, because
+      // sb.append(ch) would convert the integer to a string
+      // rather than a character.
+      char c = (char) ch;
+
+      if (!quoted) {
+        if (c == ']') {
+          break;
+        }
+        if (c == '\\') {
+          quoted = true;
+        } else {
+          if (Character.isWhitespace(c)) {
+            // The spec says "Whitespaces other than space
+            // must be converted to space".
+            sb.append(' ');
+          } else {
+            sb.append(c);
+          }
+        }
+      } else {
+        quoted = false;
+        if (Character.isWhitespace(c)) {
+          // The spec says "Any char following "\" is
+          // inserted verbatim (exception: whitespaces still
+          // have to be converted to space!)."
+          sb.append(' ');
+        } else {
+          sb.append(c);
+        }
+      }
+    }
+
+    return sb.toString();
+  }
+
+  // Parse an SGF "Text" property value.
+  private String parseComment() throws SgfError, IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.